utils/zerovec/src/varzerovec/owned.rs

```diff
@@ -115,6 +167,118 @@ impl<T: AsVarULE> VarZeroVecOwned<T> {
+        // shifting needed for the new element size.
+        let len = self.len();
+        let slice_len = self.entire_slice.len();
+
+        let prev_element = match shift_type {
+            ShiftType::Insert => {
+                let pos = self.element_position_unchecked(index);
+                // In the case of an insert, there's no previous element,
+                // so it's an empty range at the new position.
+                pos..pos
+            }
+            _ => self.element_range_unchecked(index),
+        };
+
+        // How much shifting must be done due to removal/insertion of an index.
+        let index_shift: i64 = match shift_type {
+            ShiftType::Insert => 4,
+            ShiftType::Replace => 0,
+            ShiftType::Remove => -4,
+        };
+        // The total shift in byte size of the owned slice.
+        let shift: i64 =
+            new_size as i64 - (prev_element.end - prev_element.start) as i64 + index_shift;
+        let new_slice_len = slice_len.wrapping_add(shift as usize);
+        if shift > 0 {
+            if new_slice_len > u32::MAX as usize {
+                panic!(
+                    "Attempted to grow VarZeroVec to an encoded size that does not fit within a u32"
+                );
+            }
+            self.entire_slice.reserve(shift as usize);
+        }
+
+        // Now that we've ensured there's enough space, we can shift the data around.
+        let first_affected_index = {
+            // Note: There are no references introduced between pointer creation and pointer use, and all
+            //       raw pointers are derived from a single &mut. This preserves pointer provenance.
+            let slice_range = self.entire_slice.as_mut_ptr_range();
+            let data_start = slice_range.start.add(4 + len * 4);
+            let prev_element_p =
+                data_start.add(prev_element.start)..data_start.add(prev_element.end);
+
+            let index_range = {
+                let index_start = slice_range.start.add(4 + 4 * index);
+                index_start..index_start.add(4)
+            };
+
+            if shift_type == ShiftType::Remove {
+                // Move the data before the element back by 4 to remove the index.
+                debug_assert!(prev_element_p.start >= index_range.end);
+                ptr::copy(
+                    index_range.end,
+                    index_range.start,
+                    prev_element_p.start.offset_from(index_range.end) as usize,
+                );
+            }
+
+            // Shift data after the element to its new position.
+            debug_assert!(slice_range.end >= prev_element_p.end);
+            ptr::copy(
+                prev_element_p.end,
+                prev_element_p
+                    .start
+                    .offset((new_size as i64 + index_shift) as isize),
+                slice_range.end.offset_from(prev_element_p.end) as usize,
+            );
+
+            match shift_type {
```

Comment:
thought: The actual variable this gets set to is further off, I wonder if this would help?

```suggestion
            /* let first_affected_index = */ match shift_type {
```

Reply:
I instead tried it as a delayed assignment, how's that?